tools/functional-tester/etcd-runner/main.go

```diff
@@ -212,6 +219,166 @@ func runRacer(eps []string, round int) {
 	doRounds(rcs, round)
 }
 
+func runWatcher(eps []string) {
+	var runningTime time.Duration = 30 // time for which operation should be performed
+	noOfPrefixes := 36                 // total number of prefixes which will be watched upon
+	watchPerPrefix := 10               // number of watchers per prefix
+	reqRate := 20                      // put request per second
+	keyPrePrefix := 100000             // max number of keyPrePrefixs for put operation
+
+	ctx := context.Background()
+	prefixKeysMap := make(map[string][]string) // map for holding prefixes and values
+	// pre generate prefixes equal to noOfPrefixes
+	prefixes := generateUniqueKeys(5, noOfPrefixes)
+	// pre generate keys
+	keys := generateRandomKeys(10, keyPrePrefix)
+
+	var (
+		revision int64
+		wg       sync.WaitGroup
+		pr       *clientv3.PutResponse
+		gr       *clientv3.GetResponse
+		err      error
+	)
+
+	for i := range prefixes {
+		prefix := prefixes[i]
+		// initialize prefix key map for each prefix
+		prefixKeysMap[prefix] = make([]string, 0)
```

Comment:
The whole thing you designed was overly complicated to be honest. I was trying to help you to simplify this eventually. I guess at this point, I'd provide some code to show how to get this simplified.

```go
func runWatcher(eps []string) {
	var runningTime time.Duration = 30 // time for which operation should be performed
	noOfPrefixes := 36                 // total number of prefixes which will be watched upon
	watchPerPrefix := 10               // number of watchers per prefix
	reqRate := 20                      // put request per second
	keyPrePrefix := 100000             // max number of keyPrePrefixs for put operation

	prefixes := generateUniqueKeys(5, noOfPrefixes)
	keys := generateRandomKeys(10, keyPrePrefix)

	var (
		revision int64
		wg       sync.WaitGroup
		pr       *clientv3.PutResponse
		gr       *clientv3.GetResponse
		err      error
	)

	ctx := context.Background()
	// create client for performing get and put operations
	client := randClient(eps)
	defer client.Close()

	// get revision using get request
	gr, err = client.Get(ctx, "non-existant")
	if err != nil {
		log.Fatal("Error occured while trying to get the revision.")
	}
	revision = gr.Header.Revision

	ctxt, cancel := context.WithDeadline(ctx, time.Now().Add(runningTime*time.Second))
	defer cancel()

	// generate and put keys in cluster
	limiter := rate.NewLimiter(rate.Limit(reqRate), reqRate)

	go func() {
		for i := 0; i < len(keys); i++ {
			for j := 0; j < len(prefixes); j++ {
				key := prefixes[j] + keys[i]
				// limit key put as per reqRate
				if err = limiter.Wait(ctxt); err != nil {
					break
				}

				// perform the put operation
				_, err = client.Put(ctxt, key, key)
				if err == context.DeadlineExceeded {
					break
				}

				if err != nil {
					log.Printf("Error: %v occured while trying to key: %v, value : %v to kv store.", err, key, key)
					continue
				}
			}
		}
	}()

	wg.Add(noOfPrefixes * watchPerPrefix)
	for i := 0; i < noOfPrefixes; i++ {
		for j := 0; j < watchPerPrefix; j++ {
			go func(prefix string) {
				defer wg.Done()

				rc := randClient(eps)
				defer rc.Close()

				wc := rc.Watch(ctx, prefix, clientv3.WithPrefix(), clientv3.WithRev(revision))

				for n := 0; n < len(keys); {
					select {
					case watchChan := <-wc:
						for _, event := range watchChan.Events {
							expectedKey = prefix + keys[n]
							if expectedKey != string(event.Kv.Key) {
								log.Fatalf("expected key %q, got %q", expectedKey, string(event.Kv.Key))
							}
							n++
						}
					case <-ctxt.Done():
						return
					}
				}

			}(prefixes[i])
		}
	}
	wg.Wait()
}
```

I hope this can help. All we really need to do is to make everything we put or watch pre-defined instead of trying to do some fancy stuff inside the put/watch loop.